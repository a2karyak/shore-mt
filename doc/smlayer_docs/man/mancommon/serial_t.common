.TH SERIAL_T COMMON "Jan 1999" Release \*(SV" "Shore Storage Manager" "Release 2.0" "Shore Storage Manager Reference Manual"
.so tmac.man.local
.SH NAME
serial_t \- Serial Number for Logical ID
.SH SYNOPSIS
.EX
.IN serial_t.h

struct serial_t {
    
    // The type of the only data member of serial_t is defined
    // elsewhere so that each level of the Shore software can
    // wrap the data member with its own definition, be it a 
    // class, struct, or union.
    serial_t_data data;

public:

    serial_t( bool ondisk=true);
    serial_t(uint4 start, bool remote);
    serial_t(const serial_t& s);

    // return value true indicates overflow
    bool increment(uint4 amount); // also decrements 

    bool is_remote() 	const;
    bool is_local()  	const;
    bool is_on_disk()	const;
    bool is_in_memory() const;

    bool is_null()    const;

    serial_t& operator=(const serial_t& t);

    operator==(const serial_t& s) const;
    operator!=(const serial_t& s) const;
    operator<=(const serial_t& s) const;
    operator<(const serial_t& s) const;
    operator>=(const serial_t& s) const;
    operator>(const serial_t& s) const;

    /* INPUT and OUTPUT */
    friend ostream& operator<<(ostream&, const serial_t& s);
    friend istream& operator>>(istream&, serial_t& s);

    friend istream& operator>>(istream&, serial_t_data& g);
    friend ostream& operator<<(ostream&, const serial_t_data& g);

    /* all of the following are in on-disk form: */
    static const serial_t max_local;
    static const serial_t max_remote;
    static const serial_t null;
};

.EE
.SH DESCRIPTION
.LP
Class
.FN serial_t
implements IDs that are unique to the volume containing
them.  See
.SA lid_t(common)
for a description of volume IDs and
.SA lid(ssm)
for information on how the SSM uses them.  Serial numbers are currently
4 bytes long, but we plan to make them 8 bytes long in the future.

.LP
Two bits our of each serial number are reserved for indicating
the type of the serial number.  The high-order bit indicates
if the serial number is
.I local,
indicating an intra-volume references, or 
.I remote,
indicating an inter-volume references.
The low order bit indicates if the serial number is in
on-disk form or has been
.I swizzled
(ie. converted into in-memory form).  Because of this, all
un-swizzled serial numbers (the only kind the SSM understands)
are odd numbers.

.SS Constructors

Generally, value-added server writers do not need to construct
serial numbers as this is done by SSM methods.

.LP
.B "serial_t(start, remote)
.IP
This constructor generates a serial number 
.VA start
as the serial number.  Actually, start is left-shifted one
bit and the low order bit is set to true (on-disk).
If
.VA remote
is
.FN true
the the high order bit is also set to mark the serial number
as a remote reference.

.SS Incrementing and Comparisons

.LP
.B "increment(amount)
.IP
The
.FN increment
method increments the serial number by amount (which may be
negative).  This is useful when an SSM routine returns
a consecutive range of serial numbers by specifying the starting
number and the size of the range.  To enumerate the range, simply call
.FN increment
once for each element in the range.
The return value is 
.FN true
if an overflow occurs.

.LP
The comparison operators can only be used to compare serial numbers
of the same type.  For example, with serial numbers A and B,
comparing
.C "A < B"
is incorrect if A is remote and B is local.

.LP
The
.FN is_null
method is equivalent to
.C "A == serial_t::null".
.SS Formatted I/O Methods

For 4-byte serial numbers, the input/output format is the format
for an unsigned integer.  For 8-byte serial numbers, stored
as two integers, the format is 999.999.

.SS Static Constants

There are a number of static constants.
.IP max_local 
Maximum value of a local serial number.
.IP max_remote
Maximum value of a remote serial number.
.IP null
Null is a special value often used to represent an unknown
or invalid serial number.  The default constructor creates
a serial number equivalent to null.  The null serial number is
local.

.DA

.SH SEE ALSO

.SA lid_t(common),
.SA lid(ssm).
