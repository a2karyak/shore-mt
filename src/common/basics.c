/* --------------------------------------------------------------- */
/* -- Copyright (c) 1994, 1995 Computer Sciences Department,    -- */
/* -- University of Wisconsin-Madison, subject to the terms     -- */
/* -- and conditions given in the file COPYRIGHT.  All Rights   -- */
/* -- Reserved.                                                 -- */
/* --------------------------------------------------------------- */

/*
 *  $Id: basics.c,v 1.2 1995/04/24 19:27:27 zwilling Exp $
 */
#define BASICS_C

#ifdef __GNUG__
#pragma implementation
#endif

#include <basics.h>
	
