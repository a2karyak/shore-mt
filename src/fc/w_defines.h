/*<std-header orig-src='shore' incl-file-exclusion='W_DEFINES_H' no-defines='true'>

 $Id: w_defines.h,v 1.3 2008/05/31 05:03:26 nhall Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#ifndef W_DEFINES_H
#define W_DEFINES_H

/*  -- do not edit anything above this line --   </std-header>*/

#include "shore.def"
/* shore-config.h does not have duplicate-include protection, but
   this file does, and we don't include shore-config.h anywhere else*/
#include "shore-config.h"

// OK, shore.def might or might not define SM_PAGESIZE
// If not, then shore-config.h defines it to 8K
// shore-config.h can define LARGE_PAGE based on config option --enable-largepages
// Now we need to force largepages if SM_PAGESIZE is too big
#if SM_PAGESIZE > 32768
#define LARGE_PAGE
#endif

/* Adapt shore to namespaces */
#if defined(__cplusplus) && !defined(USE_CPP_NAMESPACE_NO)
#ifdef _MSC_VER
namespace std {};
#endif
using namespace std;
#endif

/*<std-footer incl-file-exclusion='W_DEFINES_H'>  -- do not edit anything below this line -- */

#endif          /*</std-footer>*/
