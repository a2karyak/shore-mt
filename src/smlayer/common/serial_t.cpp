/*<std-header orig-src='shore'>

 $Id: serial_t.cpp,v 1.34 2001/06/23 06:06:18 bolo Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#define SERIAL_T_C

#ifdef __GNUG__
#pragma implementation "serial_t.h"
#pragma implementation "serial_t_data.h"
#endif

#include <stdlib.h>
#include <w_stream.h>
#include "basics.h"
#include "serial_t.h"
#include "dual_assert.h"

#ifdef HP_CC_BUG_1
const uint4_t serial_t::mask_remote = 0x80000000;
const uint4_t serial_t::max_inc = ((1<<overflow_shift_bits)-1);
const uint4_t serial_t::max_any = serial_t::max_inc;
const uint4_t serial_t::other_bits = ~max_inc;
#endif /* HP_CC_BUG_1 */

/*
 * Set up serial number constants
 */
const serial_t serial_t::max_local(serial_t::max_any, false);  // all bits but high
const serial_t serial_t::max_remote(serial_t::max_any, true);  // all bits set
const serial_t serial_t::null;  		  // only low bit  set

const char* serial_t::key_descr =
#   ifdef BITS64
					"u4u4";
#   else
					"u4";
#   endif /*BITS64*/


ostream& operator<<(ostream& o, const serial_t_data& g)
{
    return o 
#ifdef BITS64
	<< g._high << "."
#endif
	<< g._low;
}

istream& operator>>(istream& i, serial_t_data& g)
{
#ifdef BITS64
    char c;
#endif

    return i 
#ifdef BITS64
	>> g._high  >> c 
#endif
	>> g._low;
}

ostream& operator<<(ostream& o, const serial_t& s)
{
    return o << s.data;
}

istream& operator>>(istream& i, serial_t& s)
{
    return i >> s.data;
}

bool
serial_t::_incr(
	uint4_t		*what,
	uint4_t		amt,
	uint4_t		*overflow
)
{
	uint4_t	temp = *what;
	temp += amt;
	if( ((*overflow) = (temp & ~max_any)) ) {
		// overflow occurred

		// compiler apparently doesn't do what we would like
		// it to do with (*overflow) >>= bits
		(*overflow) = (*overflow) >> overflow_shift_bits;

		temp -= max_any;

		*what = temp;
		return 1;
	}
	*what = temp;
	return 0;
}

bool 
serial_t::increment(uint4_t amount) 
{ 
	// higher layer has to enforce this:
	dual_assert3(amount < max_inc); 

#ifdef BITS64
	bool	 was_remote = ((data._high & mask_remote)==mask_remote);
#else
	bool	 was_remote = ((data._low & mask_remote)==mask_remote);
#endif
	bool	 was_ondisk = ((data._low & mask_disk)==mask_disk);

	// don't change this if overflow occurs; use temp variables
	uint4_t	l, h, overflow;

	// turn off remote
#ifdef BITS64
	h = data._high;
	data._high &= ~mask_remote;
#else
	h = data._low;
	data._low &= ~mask_remote;
#endif

	// get low half, shifted
	l = data._low>>1;

	// increment the low half
	if( _incr(&l, amount, &overflow)) {
		// lower half overflowed

#ifndef BITS64
		goto _overflow;
#else
		// don't compile this if we only have a low half
		// h &= ~mask_remote;  this had to be done earlier
		// to make the code work with 32 bits

		if(_incr(&h, overflow, &overflow) ) {
			// the whole thing overflowed
			goto _overflow;
		}
		data._high = h;
#endif
	}

// nooverflow:
	data._low = ((l<<1)&~mask_disk);
	if(was_ondisk) {
		data._low |= mask_disk;
	}
	if(was_remote) {
#ifdef BITS64
		data._high |= mask_remote;
#else
		data._low |= mask_remote;
#endif
	}
	dual_assert3(was_remote == is_remote());
	dual_assert3(was_ondisk == is_on_disk());
	return 0;

_overflow:
	// clean up even though there was an error
	if(was_ondisk) {
		data._low |= mask_disk;
	}
	if(was_remote) {
#ifdef BITS64
		data._high |= mask_remote;
#else
		data._low |= mask_remote;
#endif
	}
	dual_assert3(was_remote == is_remote());
	dual_assert3(was_ondisk == is_on_disk());
	return 1; // caller has to handle this error
}

