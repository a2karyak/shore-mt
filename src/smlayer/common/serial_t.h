/*<std-header orig-src='shore' incl-file-exclusion='SERIAL_T_H'>

 $Id: serial_t.h,v 1.60 2001/06/26 16:48:35 bolo Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#ifndef SERIAL_T_H
#define SERIAL_T_H

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#ifdef __GNUG__
#pragma interface
#endif

/*
 * NB -- THIS FILE MUST BE LEGITIMATE INPUT TO cc and RPCGEN !!!!
 * Please do as follows:
 * a) keep all comments in traditional C style.
 * b) If you put something c++-specific make sure it's 
 * 	  got ifdefs around it
 */
#ifndef DUAL_ASSERT_H
#include "dual_assert.h"
#endif
#ifndef SERIAL_T_DATA_H
#include "serial_t_data.h"
#endif

/*
//short logical record ID (serial number)
*/
#define SERIAL_T
struct serial_t {
    
	/* The type of the only data member of serial_t is defined elsewhere */
	/* so that each level of the Shore software can wrap the data member */
	/* with its own definition, be it a class, struct, or union. 		 */

    /* NB: when serial_t expands to 8 bytes, rather than doing this:
		serial_t_data low, high
	 * do this:
		expand the type "serial_t_data"
	 */
    serial_t_data data;

#define serial_t_guts serial_t_data
#define guts data

#ifdef __cplusplus

#  	ifdef HP_CC_BUG_1
        static const uint4_t mask_remote;
	static const uint4_t max_inc;
	static const uint4_t max_any;
	static const uint4_t other_bits;
#	endif /* HP_CC_BUG_1 */

    enum {
		mask_zero = 0x0, 		   // no bits set
		mask_disk = 0x1, 		   // set --> on-disk form; this bit is 0 if
								   // in-memory reference

#  	ifndef HP_CC_BUG_1
		mask_remote = 0x80000000,  // set --> off-volume-reference
#	endif /* !HP_CC_BUG_1 */

#ifdef BITS64
#	define OVERFLOWBITS 31
#else
#	define OVERFLOWBITS 30
#endif

		overflow_shift_bits = OVERFLOWBITS
#  	ifndef HP_CC_BUG_1
		,
		max_inc = ((1<<overflow_shift_bits)-1),
		max_any = max_inc,
		other_bits = ~max_inc
#	endif /* !HP_CC_BUG_1 */
	};

// note: due to redefinition of OCRef, we should not need OBJECT_CACHE
// ifdef any more.

    /*
     * Warning: do not create a constructor that only takes one
     * basic C type as a parameter.  Otherwise, the compiler
     * may generate unwanted type casting (using temps) which can cause
     * major problems with functions having non-const reference
     * parameters of type serial_t
     */

		/* makes a null serial_t */
    serial_t( bool ondisk=true)
		{ 
#			ifdef BITS64
			data._high = 0;
#			endif
			data._low = ondisk?mask_disk:mask_zero;
		} 

	/* NB: the next constructor is for use by the storage manager 
	 * -- it shifts the given value around and sets bits indicating
	 * on/off-volume-reference, on-disk or in-memory, etc.
	 */
    serial_t(uint4_t start, bool remote)
        /* use middle 30 bits, low bit to 1, high bit to remote */
        {	
			/*
			// carefully written so that it works whether data
			// has _low only or _low and _high...
			*/
#			ifdef BITS64
			data._high = 0;
#			endif

			data._low = (start << 1) | mask_disk;

#			ifdef BITS64
			data._high |= remote ? mask_remote : mask_zero; 
#			else
			data._low |= remote ? mask_remote : mask_zero; 
#			endif
		}

    serial_t(const serial_t& s) 	{ *this = s; } 

    serial_t(const serial_t_data& g) 	{ 
		data = g; // struct copy
	} 

private:
	bool _incr(
		uint4_t		*what,
		uint4_t		amt,
		uint4_t		*overflow
	);

public:
	// return value true indicates overflow
    bool increment(uint4_t amount); // also decrements 

    bool is_remote() 	const {return (data.high() & mask_remote)==mask_remote;} 
    bool is_local()  	const {return (data.high() & mask_remote)==0; }
    bool is_on_disk()	const {return (data._low  & mask_disk)==mask_disk; }
    bool is_in_memory() const {return (data._low  & mask_disk)==0; }

	// This is a little flaky, but there *are* places where
	// we need to ignore the lowest bit (on-disk), e.g,.
	// to tell if the value is otherwise nil.
    bool equiv(const serial_t &s) const {
		// remote is never == local
		return
#ifdef BITS64
			(data._high==s.data._high) &&
#endif
		((data._low  & ~mask_disk)==(s.data._low & ~mask_disk));
	}
    bool is_null()    const { return this->equiv(serial_t::null); }

/**************  ASSIGNMENT  *****************************************/

#ifdef __GNUC__
		 // serial_t  =  serial_t 
    serial_t& operator=(const serial_t& t)    {
		data = t.data; // struct copy
        return *this;
    }
#endif
		 // serial_t  =  data 
    serial_t& operator=(const serial_t_data &g) {
	    this->data = g; // struct copy
		return  *this;
	}

	void set(const serial_t& t)
	{ data = t.data; }

	void set(const serial_t_data &g)
	{ data = g; }

/**************  COMPARISONS *****************************************/

/*  
 *  NB: if you want to compare a serial# in on-disk form
 *  with one in in-memory form, use the method equiv() rather
 *  than the operator ==
 *  (equiv ignores the lowest bit)
 */

// == 

		 // compare 2 serial_ts 
    bool operator==(const serial_t& s) const { 
		return (data._low == s.data._low)
#ifdef BITS64
			&& (data._high==s.data._high) 
#endif
		;
	}
		 // compare a serial_t  with a data 
    bool operator==(const serial_t_data& s) const { 
		return (data._low == s._low)
#ifdef BITS64
		 	&& (data._high==s._high) 
#endif
		;
	}
		// and... to make the above operator commutative:
    inline friend bool operator==(const serial_t_data& g, const serial_t&s) { 
		return s == g;
	}

// != 

		 // compare 2 serial_ts 
    bool operator!=(const serial_t& s) const { 
		return !(*this==s);
	}
		 // compare a serial_t  with a data 
    bool operator!=(const serial_t_data& s) const { 
		return !(*this==s);
	}
		// and... to make it commutative:
    inline friend bool operator!=(const serial_t_data& g, const serial_t&s) {  
		return !(s==g); 
	}

	/*
	// Noone should ever compare remote refs with local refs
	// except for equality comparisons, so we include an assertion
	// in the following operators.
	//
	// First we have operators for 2 serial_ts, then operators
	// for comparing a serial_t and a data.
	*/

// <= 
		 // compare 2 serial_ts 
    bool operator<=(const serial_t& s) const { 	
			//
			// using asserts in here creates problem with
			// HP CC when this file is run through rpcgen.
			//
			// dual_assert3(is_remote() == s.is_remote()); 
			return  
#ifdef BITS64
			(data._high < s.data._high) || ((data._high==s.data._high) && 
#endif
				(data._low <= s.data._low)
#ifdef BITS64
				)
#endif
			;
		}

		 // compare a serial_t  with a data 
    bool operator<=(const serial_t_data& g) const { 	
			return
#ifdef BITS64
			(data._high < g._high) || ((data._high==g._high) && 
#endif
				(data._low <= g._low)
#ifdef BITS64
				)
#endif
			;
		}

// < 
		 // compare 2 serial_ts 
    bool operator<(const serial_t& s) const { 
			//
			// using asserts in here creates problem with
			// HP CC when this file is used in a template
			// (at least, it happens in vas template code)
			//
			// dual_assert3(is_remote() == s.is_remote()); 
			return 
#ifdef BITS64
			(data._high < s.data._high) || ((data._high==s.data._high) && 
#endif
				(data._low < s.data._low)
#ifdef BITS64
				)
#endif
			;
		}
		 // compare a serial_t  with a data 
    bool operator<(const serial_t_data& g) const { 	
			return
#ifdef BITS64
			(data._high < g._high) || ((data._high==g._high) && 
#endif
				(data._low < g._low)
#ifdef BITS64
				)
#endif
			;
		}

// >= 
		 // compare 2 serial_ts 
    bool operator>=(const serial_t& s) const { 
			return ! ((*this) < s);
		}
		 // compare a serial_t  with a data 
    bool operator>=(const serial_t_data& g) const { 	
			return ! ((*this) < g);
		}
// > 
		 // compare 2 serial_ts 
    bool operator>(const serial_t& s) const { 
			return ! ((*this) <= s);
		}
		 // compare a serial_t  with a data 
    bool operator>(const serial_t_data& g) const { 	
			return ! ((*this <= g));
		}

		// and... to make the above 4 serial_t/data operators commutative:
    inline friend bool operator<=(const serial_t_data& g, const serial_t&s) 
		{  return (s >= g); }
    inline friend bool operator>=(const serial_t_data& g, const serial_t&s) 
		{  return (s <= g); }
    inline friend bool operator<(const serial_t_data& g, const serial_t&s) 
		{  return (s > g); }
    inline friend bool operator>(const serial_t_data& g, const serial_t&s) 
		{  return (s < g); }

/* INPUT and OUTPUT */
    friend ostream& operator<<(ostream&, const serial_t& s);
    friend istream& operator>>(istream&, serial_t& s);

    friend istream& operator>>(istream&, serial_t_data& g);
    friend ostream& operator<<(ostream&, const serial_t_data& g);

/* CONSTANTS */
    /* defined in serial_t.cpp: */
	/* all of the following are in on-disk form: */
	static const serial_t max_local;
	static const serial_t max_remote;
    static const serial_t null;

    /* this is the key descriptor for using serial_t's as btree keys */
    static const char* key_descr; 
#endif /* __cplusplus */
};

#ifdef __cplusplus
inline w_base_t::uint4_t w_hash(const serial_t& s)
{
#ifdef BITS64
	return s.data._low;  // this is reasonable since _low changes a lot
#else
	return s.data._low;
#endif
}
#endif /* __cplusplus */

/*<std-footer incl-file-exclusion='SERIAL_T_H'>  -- do not edit anything below this line -- */

#endif          /*</std-footer>*/
