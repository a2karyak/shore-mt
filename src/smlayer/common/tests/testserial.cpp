/*<std-header orig-src='shore'>

 $Id: testserial.cpp,v 1.11 1999/06/07 19:02:38 kupsch Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#undef BITS64 

#define MaxTestInc 5
#define MXDIFF (MaxTestInc+3)

#include <w_strstream.h>
#include <basics.h>
#include <serial_t.h>
#include <assert.h>

#define P(z)  pp(#z,z); 

#define PLN(z)  pp(#z,z); cout << endl;

void
pp(const char *str, const serial_t &z) 
{
	cout << "["  << str << "=" << z ;
	if(z.is_null()) {
		cout << ",null" ;
	}
	if(z.is_remote()) {
		cout << ",remote" ;
	}
	if(z.is_local()) {
		cout << ",local" ;
	}
	assert(z.is_remote() == !(z.is_local()));
	if(z.is_on_disk()) {
		cout << ",on-disk" ;
	}
	if(z.is_in_memory()) {
		cout << ",in_memory" ;
	}
	assert(z.is_in_memory() == !(z.is_on_disk()));
	cout << "]";
}

#define EQUIV(a,b)\
	if(a.equiv(b)) {\
		pp(#a,a); cout<<"equiv"; pp(#b,b);cout<<endl; \
	} else {\
		pp(#a,a); cout<<" !equiv "; pp(#b,b);cout<<endl; \
	}

#define CMP(a,op,b)\
	if(a op b) {\
		pp(#a,a); cout << #op; pp(#b,b); cout << endl;\
	} else {\
		pp(#a,a); cout << "!" << #op; pp(#b,b); cout << endl;\
	}

void
compare(
	const serial_t &a,
	const serial_t &b
)
{
	EQUIV(a,b);
	EQUIV(b,a);
	CMP(a,==,b);
	CMP(a,!=,b);
	if(a.is_remote() ) return;
	if(b.is_remote() ) return;
	CMP(a,<=,b);
	CMP(a,>=,b);
	CMP(a,>,b);
	CMP(a,<,b);
}
void
compare(
	const serial_t_data &a,
	const serial_t &b
)
{
	CMP(a,==,b);
	CMP(a,!=,b);
	if(b.is_remote() ) return;
	CMP(a,<=,b);
	CMP(a,>=,b);
	CMP(a,>,b);
	CMP(a,<,b);
}
void
compare(
	const serial_t &a,
	const serial_t_data &b
)
{
	CMP(a,==,b);
	CMP(a,!=,b);
	if(a.is_remote() ) return;

	CMP(a,<=,b);
	CMP(a,>=,b);
	CMP(a,>,b);
	CMP(a,<,b);
}
#define COMPARE(a,b) {\
	cout << "BEGIN { comparisons at "<< __LINE__ << " :" << __FILE__ << endl;\
	compare(a,b);\
	cout << "}" << endl;\
}

#ifdef _WINDOWS
unsigned long hex (long j)  { return (unsigned long) j; }
int dec (long j)  { return (int) j; }
#endif

int
main()
{
	int i;

	cout << "other_bits          " << hex((long)serial_t::other_bits) << endl;
	cout << "max_any             " << hex((long)serial_t::max_any) << endl;
	cout << "overflow_shift_bits " << dec((long)serial_t::overflow_shift_bits) << endl;
	cout << "max_inc             " << hex((long)serial_t::max_inc) << endl;
	cout << "mask_remote         " << hex((long)serial_t::mask_remote) << endl;
	cout << "expect overflow after " << dec((long)(serial_t::max_any<<1)+1) << endl;

#ifdef notdef
	serial_t xnull;

	PLN(xnull);

	serial_t xstartwith8(8,false);
	PLN(xstartwith8);
	COMPARE(xnull,xstartwith8);

	serial_t x8remote(8,true);
	PLN(x8remote);
	COMPARE(x8remote,xstartwith8);

	serial_t x8local(8,false);
	PLN(x8local);
	COMPARE(x8remote,x8local);
	COMPARE(x8local,xstartwith8);

	serial_t x8(8,false);
	serial_t x8copy(x8);
	PLN(x8);
	PLN(x8copy);
	COMPARE(x8,x8copy);
	COMPARE(x8local,x8copy);
	COMPARE(x8remote,x8copy);

	serial_t_data	g8  = {  8
#ifdef BITS64
		, 8
#endif
	};
	serial_t xg(g8);
	PLN(xg);
	COMPARE(xg,g8);
	COMPARE(g8,xg);
#endif /* notdef */

	// g-l-over should overflow into the high word but not overflow
	// completely
	static serial_t_data	glover  = { 
#ifdef BITS64
		serial_t::max_any-MXDIFF, // don't overflow the whole thing
		~0 - MXDIFF
#else
		(serial_t::max_any<<1) - MXDIFF
#endif
	};
	serial_t	lowover(glover);
	PLN(lowover);
	for(i=0; i<MaxTestInc; i++) {
		cout << "inc lowover by " << i << " ";
		if(lowover.increment((w_base_t::uint4_t)i)) {
			cout << "OVERFLOWED";
		} else {
			cout << "did not overflow";
		}
		cout << "; new value = ";
		PLN(lowover);
	}

	// g-h-over should overflow the high word 

	static serial_t_data	ghover  = {
#ifdef BITS64
		serial_t::max_any, 
		~0 - MXDIFF
#else
		(serial_t::max_any<<1) - MXDIFF
#endif
	};
	serial_t	highover(ghover);
	serial_t	hosave(highover);
	PLN(highover);
	for(i=0; i<MaxTestInc; i++) {
		cout << "inc highover by " << i << " ";
		if(highover.increment((w_base_t::uint4_t)i)) {
			cout << "OVERFLOWED";
		} else {
			cout << "no overflow";
		}
		cout << ":";
		PLN(highover);
	}
#ifdef notdef
	COMPARE(xg,glover);
	COMPARE(glover,xg);

	COMPARE(xg,ghover);
	COMPARE(ghover,xg);
#endif

	COMPARE(highover,hosave);
	COMPARE(highover,lowover);


	serial_t hi_local(serial_t::max_local);

	if( ! hi_local.increment((w_base_t::uint4_t)1)) {
		cout << "at line " << __LINE__ << " :" << __FILE__ 
		<< " ERROR - hi-local did not overflow" << endl;
	} 

	serial_t hi_remote(serial_t::max_remote);

	if( ! hi_remote.increment((w_base_t::uint4_t)1)) {
		cout << "at line " << __LINE__ << " :" << __FILE__ 
		<< " ERROR - hi-remote did not overflow" << endl;
	}

	return 0;
}

