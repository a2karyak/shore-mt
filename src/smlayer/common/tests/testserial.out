other_bits          c0000000
max_any             3fffffff
overflow_shift_bits 30
max_inc             3fffffff
mask_remote         80000000
expect overflow after 2147483647
[lowover=2147483638,local,in_memory]
inc lowover by 0 did not overflow; new value = [lowover=2147483638,local,in_memory]
inc lowover by 1 did not overflow; new value = [lowover=2147483640,local,in_memory]
inc lowover by 2 did not overflow; new value = [lowover=2147483644,local,in_memory]
inc lowover by 3 OVERFLOWED; new value = [lowover=2147483644,local,in_memory]
inc lowover by 4 OVERFLOWED; new value = [lowover=2147483644,local,in_memory]
[highover=2147483638,local,in_memory]
inc highover by 0 no overflow:[highover=2147483638,local,in_memory]
inc highover by 1 no overflow:[highover=2147483640,local,in_memory]
inc highover by 2 no overflow:[highover=2147483644,local,in_memory]
inc highover by 3 OVERFLOWED:[highover=2147483644,local,in_memory]
inc highover by 4 OVERFLOWED:[highover=2147483644,local,in_memory]
BEGIN { comparisons at 249 :testserial.cpp
[a=2147483644,local,in_memory] !equiv [b=2147483638,local,in_memory]
[b=2147483638,local,in_memory] !equiv [a=2147483644,local,in_memory]
[a=2147483644,local,in_memory]!==[b=2147483638,local,in_memory]
[a=2147483644,local,in_memory]!=[b=2147483638,local,in_memory]
[a=2147483644,local,in_memory]!<=[b=2147483638,local,in_memory]
[a=2147483644,local,in_memory]>=[b=2147483638,local,in_memory]
[a=2147483644,local,in_memory]>[b=2147483638,local,in_memory]
[a=2147483644,local,in_memory]!<[b=2147483638,local,in_memory]
}
BEGIN { comparisons at 250 :testserial.cpp
[a=2147483644,local,in_memory]equiv[b=2147483644,local,in_memory]
[b=2147483644,local,in_memory]equiv[a=2147483644,local,in_memory]
[a=2147483644,local,in_memory]==[b=2147483644,local,in_memory]
[a=2147483644,local,in_memory]!!=[b=2147483644,local,in_memory]
[a=2147483644,local,in_memory]<=[b=2147483644,local,in_memory]
[a=2147483644,local,in_memory]>=[b=2147483644,local,in_memory]
[a=2147483644,local,in_memory]!>[b=2147483644,local,in_memory]
[a=2147483644,local,in_memory]!<[b=2147483644,local,in_memory]
}
